torch/_dynamo/variables/misc.py

```diff
@@ -234,11 +258,75 @@ def visit(node):
 
         VariableTracker.apply(visit, (args, kwargs))
 
+        ctx = AutogradFunctionContextVariable.create(tx)
+        args = [ctx, *args]
+
         if requires_grad and torch.is_grad_enabled():
-            # TODO(jansel): handle this in training mode
-            unimplemented("autograd.Function with requires_grad")
+            # Note - this is the same check used in autograd/function.py, except inverted.
+            # If we want to support functorch transforms here, we will need to enable this.
+            if (
+                self.fn_cls.setup_context
+                != torch.autograd.function._SingleLevelFunction.setup_context
+            ):
+                unimplemented(
+                    "NYI - autograd.Function with custom setup_context method"
+                )
+
+            vjp_fn = self.fn_cls.vjp  # type: ignore[attr-defined]
+            if vjp_fn is not torch.autograd.Function.vjp:
+                unimplemented("NYI - User defind vjp")
+
+            from .torch import safe_or_raise_always_restore, TorchHigherOrderOperator
+
+            trampoline_autograd_apply = produce_trampoline_autograd_apply(self.fn_cls)
+            trampoline_autograd_fwd = produce_trampoline_autograd_fwd(self.fn_cls)
+            trampoline_autograd_bwd = produce_trampoline_autograd_bwd(self.fn_cls)
+
+            # NOTE [On Tracing autograd.Function w/ grad]
+            # The complex system described here revolves around the soundness evaluation of an autograd.Function in
+            # PyTorch. The system follows a well-defined strategy for tracing, which involves three key steps: tracing
+            # forward, tracing backward, and if both are sound the potential recording of an "apply" operation into the
+            # graph.We trace forward, and evaluate soundness. Soundness, in this context, refers to the absence of side
+            # effects, the avoidance of lifting new arguments into the trace, the production of a single tensor output,
+            # and a limited input scope confined to contexts, tensors, and constants. If the forward trace is sound,
+            # we install any guards accumulated from tracing. If not, we graph break. We trace backward, and evaluate
+            # for soundness, same as forward, except with more strictness. We enable a strict mode on the tx, and
+            # reject certain ops when running under this strict mode. If the backward trace is sound, we discard the
+            # trace by restoring. Otherwise, we raise.
```

Comment:
What happens to guards from the backward?

Reply:
Answered later on: we check if they are the same or not and then raise Unsupported if they aren't.